File: lib/domain/crop/entities/crop_entity.rb
# frozen_string_literal: true

module Domain
  module Crop
    module Entities
      class CropEntity
        attr_reader :id, :user_id, :name, :variety, :is_reference, :area_per_unit, :revenue_per_area, :region, :groups, :crop_stages, :associated_pests, :created_at, :updated_at

        def initialize(attributes)
          @id = attributes[:id]
          @user_id = attributes[:user_id]
          @name = attributes[:name]
          @variety = attributes[:variety]
          @is_reference = attributes[:is_reference]
          @area_per_unit = attributes[:area_per_unit]
          @revenue_per_area = attributes[:revenue_per_area]
          @region = attributes[:region]
          @groups = attributes[:groups] || []
          @crop_stages = attributes[:crop_stages] || []
          @associated_pests = attributes[:associated_pests] || []
          @created_at = attributes[:created_at]
          @updated_at = attributes[:updated_at]

          validate!
        end

        def reference?
          !!is_reference
        end

        def is_reference?
          reference?
        end

        def to_param
          id.to_s
        end

        def display_name
          [ name, variety ].compact.join(" ")
        end

        # ビュー用 alias: crop.pests で害虫一覧を取得
        def pests
          associated_pests
        end



        def as_json(options = nil)
          {
            id: id,
            user_id: user_id,
            name: name,
            variety: variety,
            is_reference: is_reference,
            area_per_unit: area_per_unit,
            revenue_per_area: revenue_per_area,
            region: region,
            groups: groups,
            associated_pests: associated_pests.map(&:to_hash),
            created_at: created_at,
            updated_at: updated_at
          }
        end

        # ハッシュからの変換（テスト用）
        def self.from_hash(hash)
          new(**hash)
        end

        private

        def validate!
          raise ArgumentError, "Name is required" if Domain::Shared.blank?(name)
        end
      end
    end
  end
end

File: crates/agrr-domain/src/crop/entities/crop_entity.rs
use crate::shared::record_ref::RecordRef;

/// Ruby: `Domain::Crop::Entities::CropEntity`
#[derive(Debug, Clone, PartialEq)]
pub struct CropEntity {
    pub id: i64,
    pub user_id: Option<i64>,
    pub name: String,
    pub variety: Option<String>,
    pub is_reference: bool,
    pub area_per_unit: Option<f64>,
    pub revenue_per_area: Option<f64>,
    pub region: Option<String>,
    pub groups: Vec<String>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
}

impl CropEntity {
    pub fn new(
        id: i64,
        name: impl Into<String>,
        user_id: Option<i64>,
        is_reference: bool,
    ) -> Result<Self, String> {
        let name = name.into();
        if name.trim().is_empty() {
            return Err("Name is required".into());
        }
        Ok(Self {
            id,
            user_id,
            name,
            variety: None,
            is_reference,
            area_per_unit: None,
            revenue_per_area: None,
            region: None,
            groups: vec![],
            created_at: None,
            updated_at: None,
        })
    }

    pub fn reference(&self) -> bool {
        self.is_reference
    }

    pub fn display_name(&self) -> String {
        match &self.variety {
            Some(v) if !v.is_empty() => format!("{} {}", self.name, v),
            _ => self.name.clone(),
        }
    }
}

impl RecordRef for CropEntity {
    fn is_reference(&self) -> bool {
        self.is_reference
    }

    fn user_id(&self) -> Option<i64> {
        self.user_id
    }
}
